applications/IgaApplication/custom_elements/shell_5p_hierarchic_element.cpp

```diff
@@ -0,0 +1,1305 @@
+
+            Hessian(2, 0) += rDDN_DDe(k, 0) * coords[2];
+            Hessian(2, 1) += rDDN_DDe(k, 2) * coords[2];
+            Hessian(2, 2) += rDDN_DDe(k, 1) * coords[2];
+        }
+    }
+
+    void Shell5pHierarchicElement::EquationIdVector(
+        EquationIdVectorType& rResult,
+        ProcessInfo& rCurrentProcessInfo
+    )
+    {
+        KRATOS_TRY
+
+        SizeType number_of_control_points = GetGeometry().size();
+
+        if (rResult.size() != 5 * number_of_control_points)
+            rResult.resize(5 * number_of_control_points, false);
+
+        for (IndexType i = 0; i < number_of_control_points; ++i) {
+            IndexType index = i * 5;
+            rResult[index]     = GetGeometry()[i].GetDof(DISPLACEMENT_X).EquationId();
+            rResult[index + 1] = GetGeometry()[i].GetDof(DISPLACEMENT_Y).EquationId();
+            rResult[index + 2] = GetGeometry()[i].GetDof(DISPLACEMENT_Z).EquationId();
+            // only ROTATION_X and ROTATION_Y used preliminarily, to avoid new declarations
+            // ROTATION_X = w_1 (first component of hierarchic shear difference vector)
+            // ROTATION_Y = w_2 (second component of hierarchic shear difference vector)
+            rResult[index + 3] = GetGeometry()[i].GetDof(ROTATION_X).EquationId();
+            rResult[index + 4] = GetGeometry()[i].GetDof(ROTATION_Y).EquationId();
+        }
+
+        KRATOS_CATCH("")
+    }
+
+    void Shell5pHierarchicElement::GetDofList(
+        DofsVectorType& rElementalDofList,
+        ProcessInfo& rCurrentProcessInfo
+    )
+    {
+        KRATOS_TRY
+
+        SizeType number_of_control_points = GetGeometry().size();
+
+        rElementalDofList.resize(0);
+        rElementalDofList.reserve(5 * number_of_control_points);
+
+        for (IndexType i = 0; i < number_of_control_points; ++i) {
+            rElementalDofList.push_back(GetGeometry()[i].pGetDof(DISPLACEMENT_X));
+            rElementalDofList.push_back(GetGeometry()[i].pGetDof(DISPLACEMENT_Y));
+            rElementalDofList.push_back(GetGeometry()[i].pGetDof(DISPLACEMENT_Z));
+            // only ROTATION_X and ROTATION_Y used preliminarily, to avoid new declarations
+            // ROTATION_X = w_1 (first component of hierarchic shear difference vector)
+            // ROTATION_Y = w_2 (second component of hierarchic shear difference vector)
+            rElementalDofList.push_back(GetGeometry()[i].pGetDof(ROTATION_X));
+            rElementalDofList.push_back(GetGeometry()[i].pGetDof(ROTATION_Y));
+        }
+
+        KRATOS_CATCH("")
+    }
+
+    int Shell5pHierarchicElement::Check(const ProcessInfo& rCurrentProcessInfo)
+    {
+        KRATOS_ERROR_IF(DISPLACEMENT.Key() == 0)
+            << "DISPLACEMENT has Key zero! check if the application is correctly registered" << std::endl;
+        KRATOS_ERROR_IF(ROTATION.Key() == 0)
+            << "ROTATION has Key zero! Shell5pHierarchic requires ROTATION dofs." << std::endl;

```

Comment:
```suggestion
```

no longer needed, now keys are assigned at compiletime and ensured to be non-zero

Reply:
cheers!